crates/rpc/src/server/accept.rs

```diff
@@ -0,0 +1,188 @@
+
+    fn layer(&self, inner: S) -> Self::Service {
+        AcceptService { inner, version: self.version.clone() }
+    }
+}
+
+/// Service responsible for handling HTTP ACCEPT headers.
+#[derive(Clone)]
+pub struct AcceptService<S> {
+    inner: S,
+    version: VersionReq,
+}
+
+impl<S, B> Service<http::Request<B>> for AcceptService<S>
+where
+    S: Service<http::Request<B>, Response = http::Response<B>> + Clone + Send + 'static,
+    S::Error: Send + 'static,
+    S::Future: Send + 'static,
+    B: Default + Send + 'static,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = BoxFuture<'static, Result<Self::Response, Self::Error>>;
+
+    fn poll_ready(&mut self, cx: &mut StdContext<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    /// Validates existence and content of HTTP ACCEPT header.
+    ///
+    /// The version specified in the value of the ACCEPT header must match
+    /// the version requirements specified by the server.
+    fn call(&mut self, request: http::Request<B>) -> Self::Future {
+        if let Some(accept_header_value) = request.headers().get(ACCEPT) {
+            // Convert header value to str.
+            match accept_header_value.to_str() {
+                // Parse the accept str.
+                Ok(accept_str) => match AcceptHeaderValue::try_from(accept_str) {
+                    // Parse the version.
+                    Ok(accept) => match Version::parse(accept.version) {
+                        // Verify the version matches configured requirements.
+                        Ok(version) => {
+                            if self.version.matches(&version) {
+                                self.inner.call(request).boxed()
+                            } else {
+                                debug!(target: COMPONENT, "Version does not match ({}/{})", version, self.version);
+                                bad_request("Client / server version mismatch").boxed()
+                            }
+                        },
+                        Err(e) => {
+                            debug!(target: COMPONENT, "Failed to parse version: {}", e);
+                            bad_request("Invalid version specified in accept header value").boxed()
+                        },
+                    },
+                    Err(e) => {
+                        debug!(target: COMPONENT, "Failed to parse accept header value: {}", e);
+                        bad_request("Invalid accept header value").boxed()
+                    },
+                },
+                Err(e) => {
+                    debug!(target: COMPONENT, "Failed to stringify accept header value: {}", e);
+                    bad_request("Invalid accept header value").boxed()
+                },
+            }
+        } else {
+            debug!(target: COMPONENT, "Request missing ACCEPT header");
+            bad_request("Missing required ACCEPT header").boxed()
+        }
```

Comment:
nit: I would maybe reduce the level of nesting here by adding a helper function which takes a `request` and returns `AcceptHeaderValue`.

Reply:
Have done with early returns instead. Think its a very readable approach